RDBCL-727 CPU credits exec

iftahbe opened the pull request.

Commit:
RDBCL-727 CPU credits exec

Commit:
RDBCL-727 skip test

ppekrol commented on src/Raven.Server/Config/Categories/ServerConfiguration.cs:
```diff
@@ -56,5 +56,27 @@ public class ServerConfiguration : ConfigurationCategory
+        [ConfigurationEntry("Server.CpuCredits.Exec.Arguments", ConfigurationEntryScope.ServerWideOnly)]
+        public string CpuCreditsExecArguments { get; set; }
+
+        [Description("EXPERT: The number of seconds between every invokation of the Cpu Credits executable. Default: 30 minutes.")]
```
typo:
- invokation  -> invocation 
- Cpu -> CPU

ppekrol commented on src/Raven.Server/Config/Categories/ServerConfiguration.cs:
```diff
@@ -56,5 +56,27 @@ public class ServerConfiguration : ConfigurationCategory
+
+        [Description("EXPERT: The number of seconds between every invokation of the Cpu Credits executable. Default: 30 minutes.")]
+        [DefaultValue(30*60)]
+        [TimeUnit(TimeUnit.Seconds)]
```
Why not in Minutes?

Better to write in description then 1800 seconds to avoid unit change

iftahbe replied:
Because in the code it is used as seconds. Do you prefer I use minutes here... and in the code do a conversion?

iftahbe replied:
Oh ok... I got what you mean now. You prefer that in the text I'll write "Default: 1800 seconds"

ppekrol replied:
Let's change this to Minutes

ppekrol commented on src/Raven.Server/Config/Categories/ServerConfiguration.cs:
```diff
@@ -56,5 +56,27 @@ public class ServerConfiguration : ConfigurationCategory
+        [Description("EXPERT: The number of seconds between every invokation of the Cpu Credits executable. Default: 30 minutes.")]
+        [DefaultValue(30*60)]
+        [TimeUnit(TimeUnit.Seconds)]
+        [ConfigurationEntry("Server.CpuCredits.Exec.SyncInterval", ConfigurationEntryScope.ServerWideOnly)]
```
SyncIntervalInSec

ppekrol commented on src/Raven.Server/Config/Categories/ServerConfiguration.cs:
```diff
@@ -56,5 +56,27 @@ public class ServerConfiguration : ConfigurationCategory
+        [ConfigurationEntry("Server.CpuCredits.Exec.SyncInterval", ConfigurationEntryScope.ServerWideOnly)]
+        public TimeSetting CpuCreditsExecSyncInterval { get; set; }
+
+        [Description("EXPERT: The number of seconds to wait for the Cpu Credits executable to exit. Default: 30 seconds")]
```
typo:
- Cpu -> CPU

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -280,12 +280,29 @@ void ConfigureKestrel(KestrelServerOptions options)
-                    Configuration.Server.CpuCreditsExhaustionBackgroundTasksThreshold != null)
+                    Configuration.Server.CpuCreditsExhaustionBackgroundTasksThreshold != null ||
+                    Configuration.Server.CpuCreditsExec != null ||
+                    Configuration.Server.CpuCreditsExecArguments != null)
```
Do we have to check arguments?

iftahbe replied:
We don't *have* to... but if one of them is defined and the others not, it will give us a change to throw an appropriate error.

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -280,12 +280,29 @@ void ConfigureKestrel(KestrelServerOptions options)
+                        CpuCreditsBalance.RemainingCpuCredits = Configuration.Server.CpuCreditsBase.Value;
+
+                        if (Logger.IsInfoEnabled)
+                            Logger.Info($"Cpu credits were configured but missing the {RavenConfiguration.GetKey(s => s.Server.CpuCreditsExec)} key.");
```
typo:
- Cpu -> CPU

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -379,20 +397,43 @@ public DynamicJsonValue ToJson()
             int remainingTimeToBackgroundAlert = 15, remainingTimeToFailvoerAlert = 5;
             AlertRaised backgroundTasksAlert = null, failoverAlert = null;
+
+            if (string.IsNullOrEmpty(Configuration.Server.CpuCreditsExec) == false)
```
Here you are checking string is null or empty, few lines up only for null. Can we check null or empty in both places?

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -379,20 +397,43 @@ public DynamicJsonValue ToJson()
+                                ? " key and the " + RavenConfiguration.GetKey(x => x.Server.CpuCreditsExecArguments)
+                                : "");
+                if (Logger.IsInfoEnabled)
+                    Logger.Operations($"Cpu credits were configured but missing the {print} key.");
```
typo:
- Cpu -> CPU

Commit:
RDBCL-727 CPU credits PR comments

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -379,20 +396,43 @@ public DynamicJsonValue ToJson()
+            }
+            else
+            {
+                var print = RavenConfiguration.GetKey(x => x.Server.CpuCreditsExec) +
```
Is it mandatory now?

iftahbe replied:
I think it should be mandatory if you supply other CPU credits configurations.

ayende replied:
so it should throw, not log

iftahbe replied:
yes

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -441,6 +481,22 @@ private void StartMonitoringCpuCredits()
+                            CpuCreditsBalance.RemainingCpuCredits = response.Remaining;
+                            CpuCreditsBalance.LastSyncTime = response.Timestamp;
+
+                            CpuCreditsBalance.History = new double[60 * 60];
```
Why create it instead of clearing it?

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -441,6 +481,22 @@ private void StartMonitoringCpuCredits()
 
                 try
                 {
+                    if (i++ == (int)Configuration.Server.CpuCreditsExecSyncInterval.AsTimeSpan.TotalSeconds)
```
Use `Stopwatch` and check for time, rather than assume that the `Delay` would be seconds

iftahbe replied:
I assumed since this is all heuristic-based, it doesn't matter if it's going to be 30 minutes or 32 minutes... Does the accuracy matter here?

ayende replied:
I don't like it because it isn't clear what you are doing.
The accuracy doesn't matter, but to understand the code you have to jump around a bit, and it is better if we said what we meant, rather than inferred it.

iftahbe replied:
ok

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -486,6 +542,87 @@ private void StartMonitoringCpuCredits()
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException($"Unable to get cpu credits by executing {command} {arguments}. Failed to start process.", e);
```
will this kill the server?

iftahbe replied:
Yes, I fixed it to be inside a try-catch with logging.

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -486,6 +542,87 @@ private void StartMonitoringCpuCredits()
+                throw new InvalidOperationException($"Unable to get cpu credits by executing {command} {arguments}. Failed to start process.", e);
+            }
+
+            using (var ms = new MemoryStream())
```
This is a lot of duplicated code.
We know that we want to change this already, so let's centralized process execution

iftahbe replied:
Will open an issue for that

Commit:
RDBCL-727 fix PR comments

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -439,6 +451,29 @@ private void StartMonitoringCpuCredits()
+                            CpuCreditsBalance.HistoryCurrentIndex = 0;
+                        }
+
+                        sw.Restart();
```
do this maybe in `finally` so if something goes wrong this will be reset and we will not execute checks too often?

iftahbe replied:
If it's in the finally we'll always restart it and not get a chance to reach 30 minutes.

Instead, I'm changing the order of execution so that restart will always come before the attempt to get credits so that if we have a failure the stopwatch will still be restarted.

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -439,6 +451,29 @@ private void StartMonitoringCpuCredits()
 
+                try
+                {
+                    if (sw.Elapsed.TotalSeconds >= (int)Configuration.Server.CpuCreditsExecSyncInterval.AsTimeSpan.TotalSeconds)
```
use TimeSpan comparison

iftahbe replied:
Why? I don't like the compare functions where you have an integer result. 
Using the >= operator does the exact same thing: https://referencesource.microsoft.com/#mscorlib/system/timespan.cs,164

Commit:
RDBCL-727 Better error handling
- fix race in test
- change sync interval to minutes
- Don't use results older than one hour
